client/components/domains/registrant-extra-info/with-contact-details-validation.jsx

```diff
@@ -32,47 +33,55 @@ export function disableSubmitButton( children ) {
 	);
 }
 
+/*
+ * Lookup error in schema to produce { path, errorMessage, errorCode } format.
+ *
+ * e.g {
+ *     path: "extra.uk.registrationNumber",
+ *     errorMessage: "A registration number is required for this registrant type.",
+ *     errorCode: "dotukRegistrantTypeRequiresRegistrationNumber",
+ * }
+ */
 export function interpretIMJVError( error, schema ) {
-	let explicitPath, errorCode;
+	let explicitPath, errorCode, errorMessage;
 
 	if ( schema ) {
 		// Search up the schema for an explicit errorField & message
 		const path = [ ...castArray( error.schemaPath ) ];
 
+		// The errorCode is primary because messages are localized, so without
+		// the code consumers couldn't do anything other than display errors.
 		do {
 			const node = get( schema, path, {} );
-			errorCode = errorCode || node.errorCode;
+			if ( ! errorCode && node.errorCode ) {
+				errorCode = node.errorCode;
+				errorMessage = node.errorMessage;
+			}
 			explicitPath = explicitPath || node.errorField;
 		} while ( path.pop() && ! ( explicitPath && errorCode ) );
 	}
 
 	// use field from error
-	const inferredPath = replace( error.field, /^data\./, '' );
+	const path = explicitPath || replace( error.field, /^data\./, '' );
 
-	return {
-		errorCode: errorCode || error.message,
-		path: explicitPath || inferredPath,
-	};
+	return { errorMessage, errorCode, path };
 }
 
 /*
- * @returns errors by field, like: { 'extra.field: name, errors: [ string ] }
+ * @returns errors by field, like: { extra: { tld: { fieldName: [ errors ] } } }
  */
 export function formatIMJVErrors( errors, schema ) {
 	return reduce(
 		errors,
-		( accumulatedErrors, error ) => {
-			// scan for errorField and errorCode at or above the failed rule
-			const details = interpretIMJVError( error, schema );
-
-			const { path, errorCode } = details;
-
-			// TODO: get localize user facing strings
-			const message = errorCode;
-
-			const previousErrorsForField = get( accumulatedErrors, path, [] );
-
-			return set( accumulatedErrors, path, [ ...previousErrorsForField, message ] );
+		( accumulatedErrors, rawError ) => {
+			// Compare the error to the schema and try to find an appropriate
+			// error message
+			const error = interpretIMJVError( rawError, schema );
+			return update(
```

Comment:
This is a nice example of currying. Is it meant to be used elsewhere, or just in `<WithContactDetailsValidation />`? It might be useful to provide an example or something else in a js doc comment to indicate whether you have to supply the remaining arguments to make it useful, particularly since it looks like it's our first ever import from `lodash/fp`.

Reply:
> it's our first ever import from lodash/fp

Heh, thanks for noticing :) Maybe I missed it, but I don't think I'm actually currying here, just enjoying that naturally spicy functional flavour :)

I'm actually using the fp version here is because it returns the whole object instead just of the updated value, which makes the code _so_ much tidier. That seems worth a comment.

I have wanted this sort of thing elsewhere, but I can't see how to extract/abstract it cleanly. It's like reducing with the `$push` command from [immutable-helper](https://github.com/kolodny/immutability-helper#available-commands), but this is my best attempt so far. If you think it's worth it (especially if you're aware of somewhere else we're doing this) I can pull it out into a util.

Reply:
> But you wanted immutability, right? So ignore me :)

No, looks like you're right, thanks! For some reason I thought that the normal versions of `set` and `update` returned the changed value and not the whole object. I do like immutability, but it doesn't make any difference in this case because we're reducing into a fresh object, so I've pulled the fp dependency and switched to vanilla :)

> For some reason the args order is different.

Yeah, lodash has been criticised for it's argument order. In fp land, you pass the function first, and in OO land, you call map on the object itself. Passing the collection first is a bit weird (e.g.   https://en.wikipedia.org/wiki/Map_(higher-order_function)#Language_comparison). This is the other reason for the fp branch, along with immutability.